packages/query-graphql/src/decorators/authorize-filter.decorator.ts

```
@@ -1,47 +1,117 @@
 import { ModifyRelationOptions } from '@nestjs-query/core';
 import { createParamDecorator, ExecutionContext } from '@nestjs/common';
 import { GqlExecutionContext } from '@nestjs/graphql';
+import { AuthorizationContext, AuthorizationOperationGroup } from '../auth';
 import { AuthorizerContext } from '../interceptors';
 
 function getContext<C>(executionContext: ExecutionContext): C {
   const gqlExecutionContext = GqlExecutionContext.create(executionContext);
   return gqlExecutionContext.getContext<C>();
 }
 
-function getAuthorizerFilter<C extends AuthorizerContext<unknown>>(context: C) {
+function getAuthorizerFilter<C extends AuthorizerContext<unknown>>(
+  context: C,
+  authorizationContext: AuthorizationContext,
+) {
   if (!context.authorizer) {
     return undefined;
   }
-  return context.authorizer.authorize(context);
+  return context.authorizer.authorize(context, authorizationContext);
 }
 
-function getRelationAuthFilter<C extends AuthorizerContext<unknown>>(context: C, relationName: string) {
+function getRelationAuthFilter<C extends AuthorizerContext<unknown>>(
+  context: C,
+  relationName: string,
+  authorizationContext: AuthorizationContext,
+) {
   if (!context.authorizer) {
     return undefined;
   }
-  return context.authorizer.authorizeRelation(relationName, context);
+  return context.authorizer.authorizeRelation(relationName, context, authorizationContext);
 }
 
-export function AuthorizerFilter<DTO>(): ParameterDecorator {
-  return createParamDecorator((data: unknown, executionContext: ExecutionContext) =>
-    getAuthorizerFilter(getContext<AuthorizerContext<DTO>>(executionContext)),
-  )();
+function getAuthorizationContext(operationNameOrContext: string | AuthorizationContext): AuthorizationContext {
```

While I like this, I'm also torn on it. I'm concerned about it relying on convention without any checks to ensure that the methodName conforms to a known convention. Without the check I think it could be easy for an implementor to add a new method that does not conform to the convention and their authorizer will not work.

I think there are a couple of paths forward.

1. Require the context to be set explicitly, this would allow for easier internal refactors when looking for usages of the `AuthorizationContext`. 
2. Throw an error if we can not determine the operation group/type. Looking at how its initialized I think throwing an error would be ok in this instance as it occurs when the decorator is run.

---

Hmm... what about backwards compatibility or if someone just wants to use `Authorizers` the basic old way?
If a user doesn't want to use the context he shouldn't be forced to pass one. In this case we could just pass `undefined` as context and keep it optional in the `Authorizer`.  We might as well log a warning in the decorator. This way it will be logged only once at startup to inform the user.

The perfect way would be to check if the `Authorizer` is using the second argument of the `authorize` function but I don't think Typescript reflection offers this possibility (at least not without using additional decorators)

---

Maybe we should just leave out the whole inference part? Just realised that the [relation resolvers](https://github.com/doug-martin/nestjs-query/blob/da60b9340879258df6b5294c3b43fff1f6bcc177/packages/query-graphql/src/resolvers/relations/update-relations.resolver.ts#L33) don't use the `Many`-prefix but the singular/plural forms of the relation DTO. So I guess the future-proof method would be to refactor the resolvers to explicitly pass the context which would also help developers looking up the source code.

---

Removed the inference part and updated resolvers in b6c7ce9.

@doug-martin : Still not sure about passing undefined / throwing error / logging warning...

---

Maybe instead of returning undefined we could return a proxy that throws an error as soon as one of the properties is accessed. Sadly this error wouldn't be thrown upfront but just-in-time when the user wants to use context-based authorization and hasn't provided any context to his custom method.

---

You can check out the proxy-solution in this [branch](https://github.com/mwoelk/nestjs-query/tree/authorizer-proxy)